### PR TITLE
add support: cross-encoder nli-MiniLM2-L6-H768 (L1/L3 evidence)

### PR DESCRIPTION
**Pull request by ssss141414**
## Summary
Adds L1 support for cross-encoder/nli-MiniLM2-L6-H768 by combining recipe updates with evaluator/resolver code updates, with highest honest goal verdict L3 PASS and overall PASS-WITH-BLOCKERS on partial coverage. The shipped behavior surfaces NLI as zero-shot-classification from metadata while preserving text-classification as the Optimum export boundary. Final CPU fp32 build/perf, L2 parity, and bounded functional smoke eval remain PASS by explicit repair impact analysis carry-forward; CPU fp16 perf remained HOST-BLOCKED and analyze remained CLI-BLOCKED due missing runtime parquet assets in repository content.

### What the model does
This is a cross-encoder NLI checkpoint that takes sentence pairs and emits contradiction/entailment/neutral logits. It is also used for zero-shot classification by converting candidate labels into NLI hypotheses and ranking by entailment behavior.

### Primary user stories
- A user supplies premise+hypothesis sentence pairs to obtain NLI label logits for entailment reasoning.
- A user supplies text and candidate labels to obtain ranked zero-shot labels from NLI scoring.

### Supported tasks
- zero-shot-classification: supported by checkpoint metadata and Transformers pipeline tag.
- text-classification: supported by Transformers auto-model class and canonical export boundary.

### Model architecture
```text
RobertaForSequenceClassification
├── RobertaModel
│   ├── Embeddings
│   └── Encoder layer x 6 (MiniLMv2-L6)
└── SequenceClassification head (3 logits: contradiction/entailment/neutral)
```

Source/confidence: frozen model-breakdown profile for this run (verified model card + config + class metadata).

### 1. Baseline
- Pinned main baseline commit: 02179d185f0e3ac23518236b5ca1f61fd37ef680.
- Baseline recipe-free config/build/perf status: PASS/PASS/PASS.
- Baseline recipe-free build summary: build complete in 40.1s (export 25.2s, optimize 13.1s), model.onnx produced.
- Baseline recipe-free perf summary: avg latency 116.18 ms, throughput 8.61 samples/s, RAM inference delta +104.2 MB.
- Starting recipe/code delta from baseline:
  - Added recipe files for CPU fp32 and fp16:
    - examples/recipes/cross-encoder_nli-MiniLM2-L6-H768/cpu/cpu/zero-shot-classification_fp32_config.json
    - examples/recipes/cross-encoder_nli-MiniLM2-L6-H768/cpu/cpu/zero-shot-classification_fp16_config.json
  - Recipe eval mapping changed to true pair semantics (input_column=premise, input_pair_column=hypothesis, label_column=label) and removed synthetic candidate_labels+hypothesis_template override path for NLI smoke.
  - Candidate branch final diff vs origin/main also includes evaluator/resolver code files and one loader test file listed in Delta below.

### 2. Goal
- Effort/Goal/Outcome: L1 / L3 / L1.
- Goal ceiling committed as L3 with acceptance criteria covering recipe-authority build/perf, parity, bounded semantic smoke eval, and analyze capture.

### 3. Outcome
- Final candidate SHA: e90e0a8cbb4445f91803942ff462ddf768a8203a.
- Highest reached goal verdict: L3 PASS.
- Overall tester verdict: PASS-WITH-BLOCKERS.
- Coverage: partial.
- REQUEST_CHANGES repair validation (r5.v1): semantic impact analysis on `27053c54..e90e0a8c` is constrained to typing/test-isolation/guard changes, so prior build/perf/L2/L3 evidence is preserved and explicitly rebound to the final SHA.
- Deferred/blocked tuples (explicit):
  - (CPUExecutionProvider, cpu, fp16, perf) = HOST-BLOCKED (bounded rerun timeout).
  - (analyze, all EP summaries requiring runtime parquet rules) = CLI-BLOCKED.
- Shipped files in branch diff vs origin/main:
  - examples/recipes/cross-encoder_nli-MiniLM2-L6-H768/cpu/cpu/zero-shot-classification_fp16_config.json
  - examples/recipes/cross-encoder_nli-MiniLM2-L6-H768/cpu/cpu/zero-shot-classification_fp32_config.json
  - src/winml/modelkit/eval/test_zero_shot_classification_evaluator.py
  - src/winml/modelkit/eval/zero_shot_classification_evaluator.py
  - src/winml/modelkit/loader/resolution.py
  - src/winml/modelkit/utils/eval_utils.py
  - tests/unit/loader/test_resolve_task_nli_zero_shot.py

### 4. Per-EP/device/precision results and Functional smoke Eval
| Tier | EP / Device | Precision | Verdict | Mean | p50 | p90 | Throughput | RAM Delta |
|---|---|---|---|---:|---:|---:|---:|---:|
| L0/L1 | CPUExecutionProvider / cpu | fp32 build | PASS | - | - | - | - | - |
| L1 | CPUExecutionProvider / cpu | fp16 build + real conversion | PASS | - | - | - | - | - |
| L1 | CPUExecutionProvider / cpu | fp32 perf | PASS | 109.04 ms | 109.28 ms | 112.66 ms | 9.17 samples/s | +104.5 MB |
| L1 | CPUExecutionProvider / cpu | fp16 perf (bounded) | HOST-BLOCKED | timeout | timeout | timeout | timeout | timeout |

- FP16 real conversion proof: PASS (float initializers 108 -> float16 initializers 108; payload bytes 328491024 -> 164249616; ratio 0.5000124934920597).
- FP16 perf bounded rerun: exit code 124 (timed out), treated as HOST-BLOCKED with explicit deferred coverage.

Functional smoke Eval (operability evidence only, not benchmark-quality):
- Runtime path: fp32 CPU artifact from final candidate lineage.
- Dataset: nyu-mll/multi_nli at revision da70db2af9d09693783c3320c4249840212ee221.
- Split: validation_matched.
- Rows requested/processed: 2/2.
- Columns mapping: input_column=premise, input_pair_column=hypothesis, label_column=label.
- Fan-out caps: candidate_label_cap=3, expanded_pair_cap_per_row=1, observed_fanout_per_row=1.
- Verified row semantics (true premise+hypothesis pairs):
  - "The new rights are nice enough" + "Everyone really likes the newest benefits "
  - "This site includes a list of all award winners and a searchable database of Government Executive articles." + "The Government Executive articles housed on the website are not able to be searched."
- Label mapping validation: dataset labels [entailment, neutral, contradiction], model order [contradiction, entailment, neutral], canonical mapping verified.
- Metrics: accuracy 1.0, F1 0.6667.
- Interpretation guardrail: these 2 rows are a bounded functional smoke only and are non-representative for model quality claims.

### 5. Delta
Recipe delta (starting and final):
- Added and populated both CPU recipes under examples/recipes/cross-encoder_nli-MiniLM2-L6-H768/cpu/cpu.
- Eval mapping now uses premise+hypothesis pair columns with label column, replacing synthetic hypothesis-template route for NLI smoke.
- Production recipe README remained untouched.

Code behavior delta:
- Metadata-derived NLI task surfacing in loader resolution:
  - NLI sequence-classification checkpoints (architecture + id2label triad) surface task zero-shot-classification.
  - Non-NLI sequence-classification siblings remain text-classification.
- Optimum export boundary preserved as text-classification.
- Evaluator adds optional columns_mapping.input_pair_column:
  - When present: evaluates true text pairs for NLI semantics.
  - When absent: preserves legacy candidate-label zero-shot behavior.

Bug fix explanation:
- Symptom/minimal trigger: prior bounded L3 eval path could produce semantically invalid NLI scoring by generating synthetic candidate-label hypotheses and not consuming dataset hypothesis text.
- Root cause: evaluator path only supported candidate-label fan-out semantics, which is valid for general zero-shot ranking but not for direct NLI pair semantics needed for MultiNLI smoke.
- Changed symbols/fix mechanism: zero-shot evaluator pair-aware logic and schema support for optional input_pair_column, plus recipe updates using premise/hypothesis/label columns.
- Generality: rule is data-driven by columns mapping and label canonicalization, with no checkpoint-name hardcoding.
- Compatibility/blast radius: legacy behavior is intentionally preserved when input_pair_column is absent; only pair-aware configs opt into new semantics.
- Regression evidence: focused pytest PASS, resolver behavior probe PASS, fp32 build/perf PASS, L2 parity PASS, bounded L3 smoke PASS with verified pair semantics.

Focused quality rerun at final SHA:
- ruff on touched files: PASS.
- pytest on evaluator+resolver tests: PASS (4 passed).
- mypy on touched production files: BASELINE-ONLY (13 errors at head, 13 at base, introduced=0).

### 6. Analyze summary - component level and op level
Status: CLI-BLOCKED. Static runtime rule analysis could not execute because runtime parquet assets are absent in tracked repository content for both candidate head and origin/main under runtime_check_rules; this is a repository content availability issue, not a candidate-code claim.

Component-level summary:
- Blocked: no analyzable component coverage rows could be emitted because analyze exited before runtime rule resolution.

Op-level summary:
- Blocked: no reliable per-op/per-EP classification summary could be emitted for the same reason.

### 7. Reproduce commands
```powershell
# Clone and checkout final candidate branch
$REPO = "<winml-cli-clone>"
Set-Location $REPO
git fetch origin
git checkout producer/nli-minilm2-r3-20260821-1
git rev-parse HEAD

# Runtime contract overlays (tester order)
$OVERLAY_BASE = "<overlay-with-transformers-5.15.0>"
$OVERLAY_EXTRA = "<extra-overlay>"
$env:PYTHONPATH = "$OVERLAY_BASE;$REPO/src;$OVERLAY_EXTRA"
$env:PYTHONNOUSERSITE = "1"
$PYTHON = "<python-3.11-embed>"

# Focused quality gates at final SHA
& $PYTHON -m ruff check src/winml/modelkit/eval/test_zero_shot_classification_evaluator.py src/winml/modelkit/eval/zero_shot_classification_evaluator.py src/winml/modelkit/loader/resolution.py tests/unit/loader/test_resolve_task_nli_zero_shot.py
& $PYTHON -m pytest src/winml/modelkit/eval/test_zero_shot_classification_evaluator.py tests/unit/loader/test_resolve_task_nli_zero_shot.py -q
& $PYTHON -m mypy --follow-imports=skip --show-error-codes --show-column-numbers --hide-error-context --no-color-output --no-error-summary --no-pretty src/winml/modelkit/eval/test_zero_shot_classification_evaluator.py src/winml/modelkit/eval/zero_shot_classification_evaluator.py src/winml/modelkit/loader/resolution.py

# Recipe-authority builds
winml build -c examples/recipes/cross-encoder_nli-MiniLM2-L6-H768/cpu/cpu/zero-shot-classification_fp32_config.json -m cross-encoder/nli-MiniLM2-L6-H768 -o <out_fp32> --ep cpu --device cpu --rebuild
winml build -c examples/recipes/cross-encoder_nli-MiniLM2-L6-H768/cpu/cpu/zero-shot-classification_fp16_config.json -m cross-encoder/nli-MiniLM2-L6-H768 -o <out_fp16> --ep cpu --device cpu --precision fp16 --rebuild

# Perf
winml perf -m <out_fp32>/model.onnx --ep cpu --device cpu --warmup 1 --iterations 3
winml perf -m <out_fp16>/model.onnx --ep cpu --device cpu --warmup 1 --iterations 3
# Bound fp16 perf to 180s and classify timeout exit 124 as HOST-BLOCKED if exceeded.

# L2 parity and bounded L3 functional smoke
# (same artifact and runtime contract as above)
winml eval -c examples/recipes/cross-encoder_nli-MiniLM2-L6-H768/cpu/cpu/zero-shot-classification_fp32_config.json -m cross-encoder/nli-MiniLM2-L6-H768 --model-path <out_fp32>/model.onnx --runtime winml --dataset nyu-mll/multi_nli --dataset-revision da70db2af9d09693783c3320c4249840212ee221 --split validation_matched --samples 2 --no-shuffle --format json

# Analyze (expected CLI-BLOCKED in current repo content)
winml analyze --model <out_fp32>/model.onnx --ep all --output <analyze_out>
```

### CODE REVIEW

**Comment by ssss141414** [comment]
### Independent Reviewer Verdict (contract)
- PR: https://github.com/microsoft/winml-cli/pull/1325
- Reviewed head SHA: `27053c54c4dd832cbdbd05e5dbfada9cb0b00318` (matches expected)
- Base: `main`
- Draft/label gate: `isDraft=true`, label `model-scale-by-skill` present
- Diff scope reviewed: 7 files (2 recipes, evaluator/resolver/schema/test files), no README churn
- Thread gate: reviewThreads enumerated, open threads = 0

### What I independently verified
- Metadata-driven NLI detection is narrow and checkpoint-name-agnostic:
  - `_is_nli_sequence_classifier` requires `*ForSequenceClassification` architecture + NLI label triad in `id2label`.
  - Non-NLI classifiers remain `text-classification`.
  - `optimum_task` remains canonical `text-classification` in the new resolver unit test.
- Pair-mode evaluator path and schema plumbing:
  - Added optional `input_pair_column` in zero-shot schema.
  - Evaluator switches to text-pair path only when `input_pair_column` is provided.
  - Legacy candidate-label path remains present and separately tested.
- Recipes are present only for CPU fp32/fp16 tuples under the expected path shape.
- Focused independent reruns from the final-verify evidence run (`tester-r4-v4-20260821-150041`):
  - `ruff-focused` exit 0
  - `pytest-focused` exit 0 (`4 passed`)
  - `mypy-head` and `mypy-base` both nonzero (baseline debt exists); classification artifact generated
- PR body hierarchy/order is correct (`Summary` -> `Model metadata` -> `Validation and support evidence`) and includes L3 caveat language plus Analyze CLI-BLOCKED disclosure.

### Blocking findings (owner-routed)
1. **[Producer] Lint/check gate is not green on reviewed SHA.**
   - GitHub `statusCheckRollup` for this SHA includes `lint = COMPLETED/FAILURE`.
   - Additional checks were still in progress at capture time (`Analyze (Python)`, `test (commands)`, `test (remaining)`).
   - Action: fix the failing lint job and re-run until the exact reviewed head SHA is fully green.

2. **[Producer] Address mypy delta indicated by independent classification artifact.**
   - The head-vs-base classifier output flags candidate-introduced diagnostics in `zero_shot_classification_evaluator.py` for duplicate local names (`predictions`/`references`) in `compute()` branching.
   - Even if this is message-shift noise, it must be resolved unambiguously in CI so lint no longer fails.

3. **[Producer] Test isolation risk in new evaluator tests.**
   - `ClassificationMetric.compute` is monkeypatched by direct assignment and not restored.
   - This can leak across tests in broader partitions depending on execution order.
   - Action: use pytest `monkeypatch` fixture/context-scoped patching to guarantee restoration.

### Non-blocking notes
- No checkpoint special-casing was found in resolver logic.
- No unrelated file churn was found in this diff.

REQUEST_CHANGES

**Comment by ssss141414** [comment]
REQUEST_CHANGES repair follow-up resolved on head e90e0a8cbb4445f91803942ff462ddf768a8203a.

- Lint/check gate: all required checks are COMPLETED/SUCCESS on this exact head.
- Mypy duplicate-local-name concern: repaired and independently revalidated with no new diagnostics versus base.
- Test isolation concern: evaluator tests now use monkeypatch-based restoration; restoration verified in independent probe evidence.

**Comment by ssss141414** [comment]
### Independent Reviewer Second Pass (contract)
- PR: https://github.com/microsoft/winml-cli/pull/1325
- Reviewed head SHA: e90e0a8cbb4445f91803942ff462ddf768a8203a
- Previous REQUEST_CHANGES comment: https://github.com/microsoft/winml-cli/pull/1325#issuecomment-5366358201
- Scope: remote PR metadata/checks + focused diff correctness + prior-issue closure verification + independent monkeypatch/label probes

### 1) GitHub checks rollup at current head
- Head SHA from current PR metadata: e90e0a8cbb4445f91803942ff462ddf768a8203a
- Completed/success checks (all green):
  - Analyze (Python)
  - lint
  - test (analyze)
  - test (models)
  - test (optim)
  - test (commands)
  - test (remaining)
  - CodeQL
  - license/cla
- Result: all listed checks are COMPLETED with SUCCESS for the reviewed head SHA.

### 2) Prior issue closure verification
- Prior blocker A (lint/check gate not green): CLOSED.
  - Current rollup is fully green at head e90e0a8cbb4445f91803942ff462ddf768a8203a.
- Prior blocker B (mypy duplicate-local/nullable diagnostics in evaluator path): CLOSED.
  - Diff now uses disambiguated locals in compute path (`pair_predictions`/`pair_references`, `candidate_predictions`/`candidate_references`).
  - Focused evaluator mypy artifact reports no issues for target evaluator file in both head/base compare context.
- Prior blocker C (test isolation via direct monkeypatch assignment): CLOSED.
  - New tests use fixture-scoped monkeypatch (`monkeypatch.setattr(...)`).
  - Independent restoration probes confirm `ClassificationMetric.compute` is restored after pytest/probe execution.

### 3) Narrow correctness review (full PR diff + behavior)
- Draft/label/base/head gate:
  - Draft: true
  - Label `model-scale-by-skill`: present
  - Base/head: `main` <- `producer/nli-minilm2-r3-20260821-1`
- No unrelated churn:
  - Diff is constrained to 7 expected files (2 recipes, evaluator, loader resolution, eval schema, 2 tests).
- NLI metadata detection safety:
  - `_is_nli_sequence_classifier` requires sequence-classification architecture plus NLI id2label triad; no checkpoint-name hardcoding.
- Legacy zero-shot path preservation:
  - Candidate-label mode is retained when `input_pair_column` is absent.
- Pair-column semantics and label mapping:
  - Pair mode enabled only when `input_pair_column` exists; canonical NLI label normalization/validation is present.
  - Independent probe confirms pair calls consume premise+hypothesis columns and legacy candidate-label behavior remains intact.
- PR body/evidence binding:
  - Body claims align with the shipped diff and the closure evidence above.

### 4) Independent evidence summary
- Checks metadata snapshot: all 9 checks success on e90e0a8cbb4445f91803942ff462ddf768a8203a.
- Independent probes:
  - monkeypatch restore probe: PASS (`compute_restored_after_pytest=true`, `compute_restored_after_probe=true`)
  - label/pair semantics probe: PASS (pair calls and mapped labels match expected triad; legacy path preserved)

No unresolved concrete defects found in this second pass.

APPROVE
